docs/src/developers/extensions.md

```diff
@@ -1,51 +1,132 @@
 ```@meta
 CurrentModule = JuMP
+DocTestSetup = quote
+    using JuMP
+end
+DocTestFilters = [r"≤|<=", r"≥|>=", r" == | = ", r" ∈ | in ", r"MathOptInterface|MOI"]
 ```
 
-Extending JuMP
-==============
+# [Extensions](@id extensions_manual)
 
-```@meta
-# TODO: How to extend JuMP: discussion on different ways to build on top of JuMP.
-# How to extend JuMP's macros and how to avoid doing this.
+JuMP provides a variety of ways to extend the basic modeling functionality.
+
+## Define a new set
+
+To define a new set for JuMP, subtype `MOI.AbstractScalarSet` or
+`MOI.AbstractVectorSet` and implement `Base.copy` for the set. That's it!
+
+```jldoctest define_new_set
+struct _NewVectorSet <: MOI.AbstractVectorSet
+    dimension::Int
+end
+Base.copy(x::_NewVectorSet) = x
+
+model = Model()
+@variable(model, x[1:2])
+@constraint(model, x in _NewVectorSet(2))
+
+# output
+
+[x[1], x[2]] ∈ _NewVectorSet(2)
 ```
 
-## Extending MOI
+However, for vector-sets, this requires the user to specify the dimension
+argument to their set, even though we could infer it from the length of `x`!
 
-```@meta
-# TODO: Create new MOI function/sets, how to use it in JuMP
+You can make a more user-friendly set by subtyping [`AbstractVectorSet`](@ref)
+and implementing [`moi_set`](@ref).
+
+```jldoctest define_new_set
+struct NewVectorSet <: JuMP.AbstractVectorSet end
+JuMP.moi_set(::NewVectorSet, dim::Int) = _NewVectorSet(dim)
+
+model = Model()
+@variable(model, x[1:2])
+@constraint(model, x in NewVectorSet())
+
+# output
+
+[x[1], x[2]] ∈ _NewVectorSet(2)
 ```
 
-### Adding a bridge
+## Extend [`@variable`](@ref)
 
-```@meta
-# TODO: create new bridge
+Just as `Bin` and `Int` create binary and integer variables, you can extend
+the [`@variable`](@ref) macro to create new types of variables. Here is an
+explanation by example, where we create a `AddTwice` type, that creates a tuple
+of two JuMP variables instead of a single varaible.
+
+First, create a new struct. This can be anything. Our struct holds a
+[`VariableInfo`](@ref) object that stores bound information, and whether the
+variable is binary or integer.
+```jldoctest new_variable
+julia> struct AddTwice
+           info::JuMP.VariableInfo
+       end
 ```
 
-See the [bridge section in the MOI manual](https://jump.dev/MathOptInterface.jl/v0.9.1/apimanual/#Automatic-reformulation-1).
+Second, implement [`build_variable`](@ref), which takes `::Type{Duplicate}` as
```

Comment:
s/Duplicate/AddTwice ?

Reply:
Good catch. I originally named the type `Duplicate`, but changed it to be more obvious.